mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,317 @@
+            return;
+        }
+
+        mlir::IRRewriter builder(func->getContext());
+
+        // 1. Identify the functions with multiple tapes
+        // Walk through each function and count the number of devices.
+        // In frontend when tracing (jax_tracers.py/trace_quantum_function),
+        // each tape has its own quantum.device operation attached to it
+
+        // Do nothing and exit for classical and single-tape programs
+        if (countTapes(func) < 2) {
+            return;
+        }
+
+        // 2. Parse the function into tapes
+        // total number of operation lists is number of tapes +2
+        // for classical pre and post processing
+        SmallVector<std::vector<Operation *>> OpsEachTape(countTapes(func) + 2,
+                                                          std::vector<Operation *>());
+
+        collectOperationsForEachTape(func, OpsEachTape);
+
+        // 3. Get the SSA values needed by the post processing (PP)
+        // These need to be returned by the tapes.
+        // Note that later tapes can also need values from earilier tapes.
+        SmallVector<Value> NecessaryValuesFromEarlierTapes;
+
+        // Go through all the operands of all the PP ops
+        // Find the ones that are not produced in PP itself
+        std::vector<Operation *> PPOps = OpsEachTape.back();
+        collectNecessaryValuesFromEarlierTapes(PPOps, NecessaryValuesFromEarlierTapes);
+
+        // 4. Generate the functions for each tape
+        unsigned int NumTapes = countTapes(func);
+        SmallVector<FailureOr<func::FuncOp>> OutlinedFuncs;
+        for (unsigned int i = 0; i < NumTapes; i++) {
+            if (failed(createTapeFunction(OpsEachTape[OpsEachTape.size() - 2 - i],
+                                          NecessaryValuesFromEarlierTapes, builder,
+                                          OpsEachTape.size() - 3 - i, func, OutlinedFuncs))) {
+                return signalPassFailure();
+            };
+        }
+
+        // OutlinedFuncs contains the tapes in reverse order (tape_2, tape_1, tape_0)
+        // Move them into the correct order
+        // Also, make the outlined functions keep the original's attributes
+        SmallVector<NamedAttribute> OutlinedFuncAttrs;
+        ArrayRef<NamedAttribute> FullOriginalFuncAttrs = func->getAttrs();
+        for (auto attr : FullOriginalFuncAttrs) {
+            StringRef attrname = attr.getName();
+            if ((attrname != "sym_name") && (attrname != "function_type")) {
+                OutlinedFuncAttrs.push_back(attr);
+            }
+        }
+        for (auto outlined : OutlinedFuncs) {
+            func::FuncOp outlinedfunc = *outlined;
+            outlinedfunc->moveAfter(func);
+            outlinedfunc->setAttrs(OutlinedFuncAttrs);
+        }
+    } // runOnOperation()
+};
+
+std::unique_ptr<Pass> createSplitMultipleTapesPass()
+{
+    return std::make_unique<SplitMultipleTapesPass>();
+}
+
+} // namespace catalyst
```

Comment:
eof newline :)))))))) (`make format` removes them)

Reply:
manually adding a eof newline makes CI formatter fail....

Reply:
Turns out the "github eof red symbol" signals that eof newline is missing, so as long as we don't see that symbol we're fine. 

`make format` is correct.